src/domain/task-system/src/services/task_scheduler.rs

```diff
@@ -16,16 +17,26 @@ use crate::*;
 /////////////////////////////////////////////////////////////////////////////////////////
 
 #[async_trait::async_trait]
-pub trait TaskService: Sync + Send {
+pub trait TaskScheduler: Sync + Send {
     /// Creates a new task from provided logical plan
     async fn create_task(&self, plan: LogicalPlan) -> Result<TaskState, CreateTaskError>;
 
-    /// Returns current state of a given task
-    async fn get_task(&self, task_id: &TaskID) -> Result<TaskState, GetTaskError>;
-
     /// Returns states of tasks associated with a given dataset ordered by
     /// creation time from newest to oldest
     fn list_tasks_by_dataset(&self, dataset_id: &DatasetID) -> TaskStateStream;
+
+    /// Returns current state of a given task
+    async fn get_task(&self, task_id: &TaskID) -> Result<TaskState, GetTaskError>;
+
+    /// Attempts to cancel the given task
+    async fn cancel_task(&self, task_id: &TaskID) -> Result<TaskState, CancelTaskError>;
+
+    /// Blocks until the next task is available for execution and takes it out
+    /// of the queue (called by [TaskExecutor])
+    async fn take(&self) -> Result<TaskID, TakeTaskError>;
+
+    /// A non-blocking version of [TaskScheduler::take()]
+    async fn try_take(&self) -> Result<Option<TaskID>, TakeTaskError>;
```

Comment:
Scheduler will maintain queues of tasks and decide when a task can run based on capacity. The executor(s) will be stealing tasks from the queues through these methods.